### ISSUE TITLE
Adding Embark to GNU ELPA

### ISSUE THREAD

**Issue by oantolin** [oantolin/embark#233]
I think it might be a good idea to add Embark to GNU ELPA at some point, so I'm opening the issue to see where we stand on FSF copyright assignment. These are the contributors to Embark that would need to sign the copyright assignment:

 - [x] @oantolin
 - [x] @minad
 - [x] @clemera
 - [ ] @a13
 - [x] @iyefrat
 - [x] @jakanakaevangeli
 - [x] @leungbk
 - [x] @jaor

To each one I ask if you have already signed the papers, and if not, if you would be willing to do so.

**Comment by leungbk** [comment]
I've signed the papers already.

**Comment by jakanakaevangeli** [comment]
Signed as Miha Rihtarsic.

**Comment by oantolin** [comment]
It's easier to do than I used to think it was, @iyefrat. In particular it does not require sending actual paper through the post any more! You email the FSF as explained below, and they will soon email you a PDF form to sign. You can print it, sign it, scan it and return it via email. After a few days, they will return the PDF with another signature and that's it.

You need to send an email to `assign@gnu.org`. The subject line should be your full legal name _if_ it can be written only using ASCII characters. If your name contains non-ASCII characters, use an ASCII transliteration _instead_ (I had to replace a "í" with "i", for example). This ASCII business is due to a quaint requirement of the U.S. copyright office! (In particular, it's not the FSF's fault.)

For the body of the email, copy the text at the end of this comment and insert your answers. The most delicate question is whether your employer has a copyright claim on your changes, you may need to consult with your employer's legal department to find out.

Email body:

```
REQUEST: SEND FORM FOR PAST AND FUTURE CHANGES

[What is the name of the program or package you're contributing to?]

GNU Emacs and packages in Emacs's GNU ELPA repository.

[Did you copy any files or text written by someone else in these changes?
Even if that material is free software, we need to know about it.]

[Do you have an employer who might have a basis to claim to own
your changes?  Do you attend a school which might make such a claim?]

[For the copyright registration, what country are you a citizen of?]

[What year were you born?]

[Please write your email address here.]

[Please write your postal address here.]

[Which files have you changed so far, and which new files have you written
so far?]

I have modified embark.el, a package which may be added to Emacs's GNU ELPA repository.
```

**Comment by iyefrat** [comment]
Thank you for the detailed explanation. I'll look into it, but it might take me a while.

**Comment by oantolin** [comment]
Sure, no problem, @iyefrat.

**Comment by jaor** [comment]
On Tue, Jul 27 2021, Omar Antolín Camarena wrote:

>   • [*] @oantolin
>   • [*] @minad
>   • [ ] @clemera
>   • [ ] @a13
>   • [ ] @iyefrat
>   • [*] @jakanakaevangeli
>   • [*] @leungbk
>   • [ ] @jaor
>
> To each one I ask if you have already signed the papers, and if not,
> if you would be willing to do so.

i signed the papers for all /emacs/ future contributions many years ago:
is that enough, or are there ELPA-specific papers? (i'd be willing if
so).

**Comment by oantolin** [comment]
Nope, you're all set, @jaor!

Thanks again for the verbose indicator. We're showing it off on reddit (I don't know if you visit the Emacs subreddit, I don't think I've seen you there.)

https://www.reddit.com/r/emacs/comments/osx5t9/recent_embark_developments_detailed_action/

**Comment by leungbk** [comment]
@iyefrat's contributions may meet the sub-15-line threshold below which signing the papers aren't necessary. A decent chunk of the changed lines correspond to some form of whitespace change.

**Comment by protesilaos** [comment]
Just to note that Clemens already has packages on GNU ELPA, such as [dired-git-info](https://git.savannah.gnu.org/cgit/emacs/elpa.git/tree/?h=externals/dired-git-info) and [frog-menu](https://git.savannah.gnu.org/cgit/emacs/elpa.git/tree/?h=externals/frog-menu).

+ [x] @oantolin
+ [x] @minad
+ [x] @clemera
+ [x] @jakanakaevangeli
+ [x] @leungbk
+ [x] @jaor
+ [ ] @a13
+ [ ] @iyefrat

**Comment by protesilaos** [comment]
The commits I could find from Itai Y. Efrat are `3ae6f51`, `a7aa68a`, `b2a9f84`.  As far as I can tell, the latter changes indentation.  Looking at the remaining two, I would say **they fall within the ~15 line limit**.  That is because it does not account for the same change over many occurrences, so an edit like this only counts once:

```diff
-    (switch-to-buffer buf)))
+    (pop-to-buffer buf)))
```

Which leaves us with two diff hunks:

1. Excerpt of commit `a7aa68a`

```diff
@@ -1776,11 +1776,13 @@ (defun embark-export-dired (files)
                    (let ((file (file-name-nondirectory path)))
                      (or (string= file ".") (string= file ".."))))
                  files)))
-  (dired (cons
-          ;; TODO: is it worth finding the deepest common containing directory?
-          (if (cl-every #'file-name-absolute-p files) "/" default-directory)
-          files))
-  (rename-buffer (format "*Embark Export Dired %s*" default-directory)))
+  (let ((buf (dired-noselect (cons
+                              ;; TODO: is it worth finding the deepest common containing directory?
+                              (if (cl-every #'file-name-absolute-p files) "/" default-directory)
+                              files))))
+    (with-current-buffer buf
+      (rename-buffer (format "*Embark Export Dired %s*" default-directory)))
+    (pop-to-buffer buf)))
 
 (autoload 'package-menu-mode "package")
```

2. And commit `3ae6f51`:

```diff
 README.org | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/README.org b/README.org
index e35ba4a..8b4bac4 100644
--- a/README.org
+++ b/README.org
@@ -420,7 +420,8 @@ **** Telling Embark about commands that prompt for tabs by name
       (let ((tab-list (or (mapcar #'(lambda (tab) (cdr (assq 'name tab)))
                                   (tab-bar-tabs))
                           (user-error "No tabs found"))))
-        (consult--read "Tabs: " tab-list
+        (consult--read tab-list
+                       :prompt "Tabs: "
                        :category 'tab))))
     (tab-bar-select-tab-by-name tab))
 #+end_src
```

So the list is updated thus:

+ [x] @oantolin
+ [x] @minad
+ [x] @clemera
+ [x] @jakanakaevangeli
+ [x] @leungbk
+ [x] @jaor
+ [x] @iyefrat
+ [ ] @a13

**Comment by protesilaos** [comment]
tl;dr: please double-check, but @a13 can have a checkbox as well.

* * *

Those must be the remaining commits.

```
eebbcf8 2020-05-20 DK: make linters happy
a35d9de 2020-05-16 DK: use when-let and alist-get
4790949 2020-05-16 DK: Add embark-keymap info
41c843f 2020-05-16 DK: Merge branch 'master' of github.com:oantolin/embark
f79e51d 2020-05-16 DK: Merge branch 'master' of github.com:oantolin/embark
7c42885 2020-05-16 DK: generate keymaps inline
34ebcb9 2020-05-16 DK: Define keymaps declaratively
```

The ones which merge from `master` cover the case where a divergent local branch from DK was pulling Omar's commits in upstream.  We can omit them from this list.

```
eebbcf8 2020-05-20 DK: make linters happy
a35d9de 2020-05-16 DK: use when-let and alist-get
4790949 2020-05-16 DK: Add embark-keymap info
7c42885 2020-05-16 DK: generate keymaps inline
34ebcb9 2020-05-16 DK: Define keymaps declaratively
```

A quick check:

+ `eebbcf8` contains two typos and an indentation change.  Even they are still applicable, they are a negligible change for our legal purposes.
+ `a35d9de` makes some changes to `embark-act`.  I would expect those to no longer apply, given that this command has changed a lot over time.  `git blame` tells me that nothing of that remains, but please double check.
+ `4790949` adds three lines of docs and a small code sample to the README.  A quick look suggests that it no longer exists.
+ `7c42885` changes the old implementation of keymaps and affects the `embark-keymap` function.  That function no longer exists.  It has been replaced by `embark-keymap-help` and `embark-keymap-prompter`.  Again, `git blame` tells me that no old code remains.
+ `34ebcb9` is also about the old keymap implementation.  Those have undergone thoroughgoing reform since last year.

In conclusion: what remains from DK's contributions does not block the
process.  Please verify those findings, otherwise embark is eligible for
inclusion in GNU ELPA.

And I claim that this does not constitute a blocking issue, because I have read before on emacs-devel (I think by Richard Stallman?) that in cases where we cannot secure the paperwork, the relevant code needs to be written anew.  In other words, Git history does not count, but only the current files.  Though please confirm this with the Emacs maintainers.

**Comment by iyefrat** [comment]
Hey! Thank you @protesilaos for the thorough review. I think I can verify this independently:
```bash
embark on master
❯ for file in $(git ls-files); do git blame -w -f  $file | grep "Itai"; done
3ae6f518 README.org (Itai Y. Efrat          2021-05-20 20:02:04 +0300 440)         (consult--read tab-list
3ae6f518 README.org (Itai Y. Efrat          2021-05-20 20:02:04 +0300 441)                        :prompt "Tabs: "
a7aa68af embark-consult.el         (Itai Y. Efrat          2021-05-16 13:31:29 +0300 150)     (pop-to-buffer buf)))
a7aa68af embark-consult.el         (Itai Y. Efrat          2021-05-16 13:31:29 +0300 172)     (pop-to-buffer buf)))
a7aa68af embark.el (Itai Y. Efrat          2021-05-16 13:31:29 +0300 2487)            files))))
a7aa68af embark.el (Itai Y. Efrat          2021-05-16 13:31:29 +0300 2488)     (with-current-buffer buf
a7aa68af embark.el (Itai Y. Efrat          2021-05-16 13:31:29 +0300 2490)     (pop-to-buffer buf)))
a7aa68af embark.el (Itai Y. Efrat          2021-05-16 13:31:29 +0300 2501)     (pop-to-buffer buf)))
embark on master
❯ for file in $(git ls-files); do git blame -w $file | grep "Itai"; done | wc -l
8
embark on master
embark on master
❯ for file in $(git ls-files); do git blame -w -f $file | grep "DK"; done
47909497 README.org (DK                     2020-05-16 21:06:54 +0300 358)
47909497 README.org (DK                     2020-05-16 21:06:54 +0300 359) #+BEGIN_SRC emacs-lisp
47909497 README.org (DK                     2020-05-16 21:06:54 +0300 365) #+END_SRC
47909497 README.org (DK                     2020-05-16 21:06:54 +0300 366)
a35d9de4 embark.el (DK                     2020-05-16 22:32:26 +0300  109)
a35d9de4 embark.el (DK                     2020-05-16 22:32:26 +0300  689)         (user-error "No completion here")
34ebcb99 embark.el (DK                     2020-05-16 15:53:27 +0300 2665)
34ebcb99 embark.el (DK                     2020-05-16 15:53:27 +0300 2830)
embark on master
❯ for file in $(git ls-files); do git blame -w -f $file | grep "DK"; done | wc -l
8
```
Given this, I don't think either of us are merge blockers. Unfortunately getting my fsf copyright sorted is taking longer than expected, I'll update ya'll when I have news on that front. I'd like to make more substantial contributions the the packages themselves once I'm able to.

**Comment by minad** [comment]
Omar, given Prot's analysis what do you think about sending the package to ELPA?

**Comment by oantolin** [comment]
> Omar, given Prot's analysis what do you think about sending the package to ELPA?

I guess it's bound to happen sooner or later, might as well get it over with.

**Comment by phikal** [comment]
Has there been any progress on this issue? It appears that if @protesilaos is right, the packages could be added right now? If you want to, I could test and prepare a proposal on emacs-devel.

**Comment by legendre6891** [comment]
Want to bump this issue again; `embark` is a truly wonderful package, and it would be great to have it in the OOTB experience for Emacs. Thanks all! (cc @oantolin)

**Comment by minad** [comment]
@iyefrat also got the assignment btw. The only commits in question are by @a13, but according to Prot's analysis we are alright. I hope if @oantolin gives his okay, @phikal will prepare the proposal on emacs-devel and add it to the ELPA archive with auto-sync. Nothing changes for the development here.

**Comment by oantolin** [comment]
To be perfectly honest the main reason I haven't done this is that I've had some slightly unpleasant interactions with the mailing list. Nothing truly bad, but annoying enough that I keep putting this off (well, that and the fact that I still think there is much benefit to be gained from having the package in GNU ELPA given it is already on MELPA). If @phikal is kind enough to handle the proposal, I'd be very happy to go ahead with this.

And even if I had to do it myself, I would but I'd be in no hurry to do it.

**Comment by phikal** [comment]
Done: https://mail.gnu.org/archive/html/emacs-devel/2021-10/msg02338.html

The only question is if you want Embark to be one package, along with avy-embark-collect and embark-consult, or should they be separated like on MELPA?

**Comment by oantolin** [comment]
Thanks so much!

**Comment by oantolin** [comment]
As for the packaging, I have no strong preference. Maybe I'd be inclined to keep them separate because they have differing dependencies. Do you (or anyone else) have any suggestion about this?

**Comment by phikal** [comment]
Thinking about it, it might be better to keep them in separate packages, instead of introducing "soft" dependencies.

**Comment by minad** [comment]
@oantolin 

> To be perfectly honest the main reason I haven't done this is that I've had some slightly unpleasant interactions with the mailing list. Nothing truly bad...

This is a common pattern. My interactions with emacs-devel have been *truly bad* and since then I've ceased all my direct interactions with emacs-devel. There are individuals acting in a collaborative way like @monnier or @phikal in this particular occasion. I am glad that @phikal is taking this proxy role here.

@phikal 

> Thinking about it, it might be better to keep them in separate packages, instead of introducing "soft" dependencies.

As a contributor to Embark (and maintainer of Consult, which is affected here by embark-consult) I'd like to insist to *not* introduce a soft dependency. Please keep the packaging as it is on MELPA. This is crucial to ensure bytecode compilation without warnings. In the involved packages of this completion ecosystem we tried to keep quite a strict discipline regarding to how package dependencies are specified and enforced.

**Comment by minad** [comment]
@phikal @oantolin Is it necessary to even package to avy-embark? As far as I understood you intend to deprecate this package? We could leave it on MELPA and let it rot there, or even remove it altogether?

**Comment by oantolin** [comment]
Good point about avy-embark-collect. I don't use it personally anymore. I don't know if anyone uses it, I don't think I've ever received any bug reports or feature requests for it either. It can easily be replaced by link-hint, which is more general.

Its probably best to not add it to GNU ELPA, though I can easily just keep it in maintenance mode, just test every now and then as Embark and Avy change.

**Comment by phikal** [comment]
Stefan Monnier ack'ed the request, which is enough for me. I can add the package to GNU ELPA straight away. The package won't be published before the [copyright line](https://github.com/oantolin/embark/blob/master/embark.el#L3) is updated and the version attribute is bumped. From my local testing that should be the only remaining change.

**Comment by iyefrat** [comment]
> @iyefrat also got the assignment btw.

To be clear, my assignment explicitly doesn't cover my past embark contributions since I'm not sure if I or the institution I was under at the time own the copyright. But since it ended up being like 8 lines, it shouldn't matter either way. It does mean that I can make future contributions without issue though.

**Comment by oantolin** [comment]
I changed the copyright lines and bumped the version numbers (bc3e4654329563ab28a96003610634c3a5e5a484)!

**Comment by dklimov-allstreet** [comment]
> tl;dr: please double-check, but @a13 can have a checkbox as well.
> 
> Those must be the remaining commits.
> 
> 
> ```
> eebbcf8 2020-05-20 DK: make linters happy
> a35d9de 2020-05-16 DK: use when-let and alist-get
> 4790949 2020-05-16 DK: Add embark-keymap info
> 41c843f 2020-05-16 DK: Merge branch 'master' of github.com:oantolin/embark
> f79e51d 2020-05-16 DK: Merge branch 'master' of github.com:oantolin/embark
> 7c42885 2020-05-16 DK: generate keymaps inline
> 34ebcb9 2020-05-16 DK: Define keymaps declaratively
> ```
> 
> The ones which merge from `master` cover the case where a divergent local branch from DK was pulling Omar's commits in upstream. We can omit them from this list.
> 
> 
> ```
> eebbcf8 2020-05-20 DK: make linters happy
> a35d9de 2020-05-16 DK: use when-let and alist-get
> 4790949 2020-05-16 DK: Add embark-keymap info
> 7c42885 2020-05-16 DK: generate keymaps inline
> 34ebcb9 2020-05-16 DK: Define keymaps declaratively
> ```
> 
> A quick check:
> 
> * `eebbcf8` contains two typos and an indentation change.  Even they are still applicable, they are a negligible change for our legal purposes.
> * `a35d9de` makes some changes to `embark-act`.  I would expect those to no longer apply, given that this command has changed a lot over time.  `git blame` tells me that nothing of that remains, but please double check.
> * `4790949` adds three lines of docs and a small code sample to the README.  A quick look suggests that it no longer exists.
> * `7c42885` changes the old implementation of keymaps and affects the `embark-keymap` function.  That function no longer exists.  It has been replaced by `embark-keymap-help` and `embark-keymap-prompter`.  Again, `git blame` tells me that no old code remains.
> * `34ebcb9` is also about the old keymap implementation.  Those have undergone thoroughgoing reform since last year.
> 
> In conclusion: what remains from DK's contributions does not block the process. Please verify those findings, otherwise embark is eligible for inclusion in GNU ELPA.
> 
> And I claim that this does not constitute a blocking issue, because I have read before on emacs-devel (I think by Richard Stallman?) that in cases where we cannot secure the paperwork, the relevant code needs to be written anew. In other words, Git history does not count, but only the current files. Though please confirm this with the Emacs maintainers.


Dear Mr. Stavrou, could you please tell Mr. John Wiegley that he could resolve the paper issue with `use-package` https://github.com/jwiegley/use-package/issues/282 many years ago? Just do some refactoring and don't care anymore!

**Comment by a13** [comment]
> And I claim that this does not constitute a blocking issue, because I have read before on emacs-devel (I think by Richard Stallman?) that in cases where we cannot secure the paperwork, the relevant code needs to be written anew. 

> In other words, Git history does not count, but only the current files. Though please confirm this with the Emacs maintainers.

Dear Mr. Stavrou, could you please tell Mr. John Wiegley that he could resolve the paper issue with use-package jwiegley/use-package#282 many years ago? Just do some refactoring and don't care anymore! 

Nicely done! (sarcasm)

**Comment by minad** [comment]
@a13 I know you are being sarcastic. But I'd like to be very clear that this is not how it works. If you do only refactoring, the derived code inherits the copyright. I'd say the quote by @protesilaos "In other words, Git history does not count, but only the current files." is incorrect in full generality. However it does not affect the outcome in this particular case and this is what matters. Legal matters are usually not black and white. To confirm @protesilaos' analysis once more:

* a35d9de makes minor changes to `embark-act` (use `when-let` and `alist-get`). The command `embark-act` has been rewritten multiple times in the meantime. The character of the `when-let`/`alist-get` change could not have meaningfully affected the outcome.
* The commits 7c42885 and 34ebcb9 combined introduce the `embark-keymap` helper and shuffle around the code to use this helper. In the meantime the helper has been replaced by a macro `embark-define-keymap` and the keymaps themselves have been changed again. But I think it is fair to say that parts of your code live on in `embark-define-keymap`. For reference the function you've introduced had 11 lines and looked like this:

~~~ elisp
(defun embark-keymap (keymap-alist &optional parent-map)
  "Generage keymap based on KEYMAP-ALIST if PARENT-MAP is non-nil, set it as the parent."
  (let ((map (make-sparse-keymap)))
    (mapc (lambda (keypair)
            (pcase keypair
              (`(,key . ,fn)
               (define-key map (kbd key) fn))))
          keymap-alist)
    (when parent-map
      (set-keymap-parent map parent-map))
    map))
~~~

As of now the macro looks like this:

~~~ elisp
(defmacro embark-define-keymap (name doc &rest bindings)
  "Define keymap variable NAME.
DOC is the documentation string.

BINDINGS specifies the key bindings as (string command) pairs;
the strings are passed to `kbd' to determine which key sequence
to bind.

Before the actual list of binding pairs you can include the
keyword `:parent' followed by a keymap, to specify a parent for
the defined keymap.  If the `:parent' keymap is absent,
`embark-general-map' is used by default."
  (declare (indent 1))
  (let* ((map (make-symbol "map"))
         (parent (if (eq :parent (car bindings))
                     (cadr bindings)
                   'embark-general-map))
         (bindings (if (eq :parent (car bindings)) (cddr bindings) bindings)))
    `(defvar ,name
       (let ((,map (make-sparse-keymap)))
         ,@(mapcar (pcase-lambda (`(,key ,fn))
                     (when (stringp key) (setq key (kbd key)))
                     `(define-key ,map ,key ,(if (symbolp fn) `#',fn fn)))
                   bindings)
         ,(if parent `(make-composed-keymap ,map ,parent) map))
       ,doc)))
~~~

**Comment by protesilaos** [comment]
> But I'd like to be very clear that this is not how it works. If you do
> only refactoring, the derived code inherits the copyright. I'd say the
> quote by @protesilaos "In other words, Git history does not count, but
> only the current files." is incorrect in full generality.

This is correct and is why those things are tricky.  In the case of
Embark, where (i) its history is fairly short, (ii) the original work
specific in scope, and (iii) the path of modifications straightforward,
we can be confident that our interpretation is correct.  But for
use-package I would not be so sure: there are years worth of history
there that need to be carefully assessed by those who know the
project---I don't (and I understand why it is taking them years to
finalise the process).

With regard to derivatives in general, I feel we are entering into a
gray area when some code "inspires" other code, which then engenders a
complete rewrite, only to be followed by a third refactoring, and so on.
At what point does the original work stop being the dominant influence?
Put differently, does the copyright claim of the original prevail over
those of the rewrites and if so, how levels deep?  And if yes, what are
the degrees involved and do they play any significance in the evaluation
(e.g. 10 lines of original work which over the course of recursion
derive 1000 lines---just an extreme example to make sense of things)?
We need legal experts for those issues and is why copyright assignment
to the FSF is no trivial matter.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by a13** [comment]
> 7c42885 and 34ebcb9

@minad ok, let's see how the part of code changed

mapc -> mapcar
lambda+pcase -> pcase-lambda
set-keymap-parent -> make-composed-keymap 

to me it looks like pretty straightforward refactoring

**Comment by minad** [comment]
@a13 

> to me it looks like pretty straightforward refactoring

I've not said otherwise. Please read what I wrote above: "But I think it is fair to say that parts of your code live on in embark-define-keymap." But this is beside the point - the point is that the remnants of your contributions to Embark do not surpass a limit of 15 lines at which they would become relevant for the copyright assignment process. This is the result of the contribution analysis by @protesilaos. Given the short history of Embark it was possible to trace how your contributions affected (or in fact did not affect) other code. In contrast, as @protesilaos already pointed out, such an analysis is unfeasible for older projects with many more contributors like use-package.

**Comment by a13** [comment]
> This is the result of the contribution analysis by @protesilaos.

Actually the result of the analysis is:

> 7c42885 changes the old implementation of keymaps and affects the embark-keymap function. **That function no longer exists. It has been replaced by embark-keymap-help and embark-keymap-prompter**. Again, **git blame** tells me that no old code remains.
> 34ebcb9 is also about the old keymap implementation. Those have undergone thoroughgoing reform since last year.

which sounds a bit different from 

> But this is beside the point - the point is that the remnants of your contributions to Embark do not surpass a limit of 15 lines at which they would become relevant for the copyright assignment process.

So while I completely agree with the latter passage, the so called "analysis" doesn't look that good to me.

**Comment by monnier** [comment]
> But for use-package I would not be so sure: there are years worth of
> history there that need to be carefully assessed by those who know the
> project---I don't (and I understand why it is taking them years to
> finalise the process).

[ Side note ]

I thought the copyright issue for use-package had been resolved (and
what remains is only to decide how to integrate the package into Emacs,
since John wants to bundle with Emacs rather than just add it as-is to
GNU ELPA).

**Comment by a13** [comment]
> since John wants to bundle with Emacs rather than just add it as-is to
GNU ELPA

And it makes sense, having it on ELPA doesn't free you from boostrapping. Which makes the pre-use-package part of your config not so tidy and inconsistent with the rest.

**Comment by minad** [comment]
@a13 To me it is not entirely clear what you are claiming with respect to Embark. The analysis by @protesilaos may have some minor imprecisions but the overall result has been confirmed independently multiple times, for example by me in https://github.com/oantolin/embark/issues/233#issuecomment-957116642. The outcome is that there is a single macro which was affected by your contributions: `embark-define-keymap`. You originally contributed the function `embark-keymap`, which was later refactored to the aforementioned macro. Your function `embark-keymap` had 11 lines of code, which fall below a critical threshold, where copyright claims can be reasonably made. Do you disagree with that?

**Comment by a13** [comment]
>The outcome is that there is a single macro which was affected by your contributions: embark-define-keymap

don't forget the idea to not write the walls of `define-key` manually 

> Do you disagree with that?

I absolutely agree

> may have some minor imprecisions

I wouldn't call using `git blame` as a measure of contribution in that case (where legal stuff involved) "a minor imprecision".

**Comment by minad** [comment]
> don't forget the idea to not write the walls of define-key manually

Which is hardly an original idea, if you look at your beloved use-package. However I certainly see the problem: if people have ideas, someone else implements them, who owns the copyright? For example if we look at embark-define-keymap, I could easily specify the macro in plain words without giving the "algorithm" and someone else implements it for me according to the specification. This would be a clean room implementation, which would be unencumbered by copyright. However the problem may be even harder than one thinks. Since @monnier is around in this thread - what if I give the specification in a specification language so precise that the actual code follows exactly by proof search? Who will then own the copyright? My guess is that the one who performs the (nontrivial) translation to the specification language owns the copyright. However one might actually consider the translation process to be some kind of refactoring to the type level.

> I absolutely agree

Good.

> I wouldn't call using git blame as a measure of contribution in that case (where legal stuff involved) "a minor imprecision".

This is not what was done. Your contributions have been traced to determine which parts of the code have been affected by it. I know that I have done this after @protesilaos made his initial analysis here, but I haven't spoken up about it. And after you turned up here, I just redid the same tracing steps. As I said, the outcome matters.

**Comment by protesilaos** [comment]
>>The outcome is that there is a single macro which was affected by your
>>contributions: embark-define-keymap
>
> don't forget the idea to not write the walls of `define-key` manually
>
>> Do you disagree with that?
>
> I absolutely agree
>
>> may have some minor imprecisions
>
> I wouldn't call using `git blame` as a measure of contribution in that
> case (where legal stuff involved) "a minor imprecision".

Can you please state your case explicitly so that we don't keep querying
you for one hint at a time?  What is, in your opinion, the right method?
And, when applied, what are the findings of said method?

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by monnier** [comment]
> Which is hardly an original idea, if you look at your beloved
> use-package. However I certainly see the problem: if people have ideas,
> someone else implements them, who owns the copyright?

Copyright does not apply to ideas but only to expressions of them.

**Comment by a13** [comment]
> And after you turned up here, I just redid the same tracing steps

> Which is hardly an original idea

sure, but no one used it in embark before me 

> I could easily specify the macro in plain words without giving the "algorithm" and someone else implements it for me according to the specification. 

Another problem is: where's the borderline between "plain words" and "algorithm"? Is "given a keymap and a list of key/command pairs, bind every key to a corresponding command in the keymap" an algorithm already?

> what if I give the specification in a specification language

there are languages that already look almost like specification ones (on my dayjob I constantly find myself literally writing what's written in specs), I wish Emacs Lisp was among them too

> This is not what was done.

this is what was mentioned

> And after you turned up here, I just redid the same tracing steps

**Comment by a13** [comment]
> Copyright does not apply to ideas but only to expressions of them.

repeating myself: Is "given a keymap and a list of key/command pairs, bind every key to a corresponding command in the keymap" an algorithm already?

**Comment by monnier** [comment]
> Can you please state your case explicitly so that we don't keep querying
> you for one hint at a time?  What is, in your opinion, the right method?
> And, when applied, what are the findings of said method?

I think more importantly, I'd like to know if it means you want us to
remove your code because you don't want it distributed from GNU ELPA.

AFAIK we're perfectly allowed to distribute it (to the extent that your
contributions are presumably done under the GPL or a compatible license,
otherwise you'd have complained earlier already), but I'd rather we
don't distribute code against the author's explicit wish.

**Comment by a13** [comment]
>  I'd like to know if it means you want us to
remove your code because you don't want it distributed from GNU ELPA.

no, it doesn't mean that.

**Comment by oantolin** [comment]
Hi @a13. This situation was a bit unfortunate and I apologize. I'm a little uncertain about what you want done and would love some clarification. I'm glad there is a less noisy way to declare the keymaps in the Embark source code, but I definitely don't want to use that idea against your wishes! I can go back to the wall of `define-key` if you prefer. I would of course be much happier if you signed the FSF copyright assignment and contributed again to Embark in the future, but if you don't wish to sign I totally respect that. I myself didn't sign because I really *wanted* to, but merely because I ---and my employer!--- *didn't mind* signing and other people seemed to want Embark on GNU ELPA.

**Comment by a13** [comment]
> This situation was a bit unfortunate and I apologize.

it's not your fault

> I'm a little uncertain about what you want done and would love some clarification.

I don't want 

> I'm glad there is a less noisy way to declare the keymaps in the Embark source code, but I definitely don't want to use that idea against your wishes! I can go back to the wall of define-key if you prefer.

I've already answered to similar questions twice in this thread: I'm fine with the status quo.

> I would of course be much happier if you signed the FSF copyright assignment and contributed again to Embark in the future, but if you don't wish to sign I totally respect that.

Thank you, I'm afraid it's not gonna happen any time soon. There are several reasons, I hope @bbatsov have nothing against me citing him on the subject:

> When I was approached about the possibility of including Projectile in ELPA or Emacs the project already had a lot of contributors besides me, which meant chasing a lot of people about the FSF paperwork. **More importantly, however, it meant that likely the project was going to see significantly fewer contributions down the road, as people generally don't want to be bothered with paperwork**. I hope that down the road the bar to Emacs contributions will be lowered, but as it stands today - it is too big of a trade-off for me.

Besides that, the more I live - the less I see the point of the assignment. Most GPLed software exist without it and don't face any problems related to copyright. Those GPL violation cases became known to the public (so FSF lawyers could in theory help) are just drops in the ocean (see gpl-violations.org), and they are definitely not Emacs packages :).

> > I myself didn't sign because I really wanted to, but merely because I (and my employer!) didn't mind signing

good for you

> and other people seemed to want Embark on GNU ELPA.

I personally see no reason to, especially when there's Non-GNU ELPA

**Comment by phikal** [comment]
"D.K." ***@***.***> writes:

>> and other people seemed to want Embark on GNU ELPA.
>
> I personally see no reason to, especially when there's Non-GNU ELPA

NonGNU ELPA is better than nothing, but GNU ELPA packages are in a
better position to share their code with core Emacs (and vice versa).
I think that especially packages like Embark might have potential on
that front.

-- 
	Philip Kaludercic

**Comment by minad** [comment]
@a13

> Those GPL violation cases became known to the public (so FSF lawyers could in theory help) are just drops in the ocean (see gpl-violations.org), and they are definitely not Emacs packages :).

You see - this is because Emacs is so well protected ;)

**Comment by a13** [comment]
> NonGNU ELPA is better than nothing, but GNU ELPA packages are in a
better position to share their code with core Emacs (and vice versa).

better for what? I can reuse Emacs code in any of my packages anyway, and any Emacs user can use my packages.What's the difference?

**Comment by a13** [comment]
> You see - this is because Emacs is so well protected ;)

_it reminds me an old Russian joke about a cowboy named Uncatchable Joe_

**Comment by minad** [comment]
@a13 Now you've made me read the Wikipedia article about [Russian jokes](https://en.m.wikipedia.org/wiki/Russian_jokes).

> A New Russian enters musical instruments store and exits with a drum. Boasting to his friends, he tells them: "I just paid million bucks for this, the salesman told me it's a Stradivarius drum!" His friends laugh at him, saying: "You dumb-ass, Stradivarius built only string instruments..." The New Russian rushes back into the store with a murderous look in his eyes, but exits peacefully and tells them: "Nah, it's alright... The salesman told me that Stradivarius built string instruments only for dorks, for right fellas like us he built drums".

> A lecturer visits the mental hospital and gives a lecture about how great communism is. Everybody claps loudly except for one person who keeps quiet. The lecturer asks: "Why aren't you clapping?" and the person replies "I'm not a psycho, I just work here."

> A grandson asks his grandfather: "Grandpa, is it true that in 1986 there was an accident at Chernobyl Nuclear Power Plant?" / "Yes, there was", answers the Grandpa and patted the grandson's head. / "Grandpa, is it true that it had absolutely no consequences?" / "Yes, absolutely", answered the Grandpa, and patted the grandson's second head.
